tools/lkl/lib/virtio_net.c

```diff
@@ -66,32 +66,55 @@ static inline int is_tx_queue(struct virtio_dev *dev, struct virtio_queue *queue
        return &dev->queue[TX_QUEUE_IDX] == queue;
 }
 
-static int net_enqueue(struct virtio_dev *dev, struct virtio_req *req)
+/* Inside the LKL kernel, we wrapped this buffer (which corresponds to
+ * an IP packet) with a header to make sure it got routed to the TAP
+ * device. This header is meaningless to the oustide world, so we cut
+ * it off the top of the packet before sending it to the TAP
+ * device. */
+void strip_dev_buf_header(const struct lkl_dev_buf *in_buf,
+			struct lkl_dev_buf *out_buf,
+			struct lkl_virtio_net_hdr_v1 **out_hdr)
 {
-	struct lkl_virtio_net_hdr_v1 *header;
-	struct virtio_net_dev *net_dev;
-	int ret, len;
-	void *buf;
+	memset(out_buf, 0, sizeof(*out_buf));
 
-	header = req->buf[0].addr;
-	net_dev = netdev_of(dev);
-	len = req->buf[0].len - sizeof(*header);
+	/* Header starts at the top of the buffer */
+	*out_hdr = in_buf->addr;
+
+	out_buf->len = in_buf->len - sizeof(**out_hdr);
+	out_buf->addr = in_buf->addr + sizeof(**out_hdr);
+}
 
-	buf = &header[1];
+static int net_enqueue(struct virtio_dev *dev, struct virtio_req *req)
+{
+	struct lkl_virtio_net_hdr_v1 *header;
+	struct lkl_dev_buf out_buf;
+	int ret;
 
-	if (!len && req->buf_count > 1) {
-		buf = req->buf[1].addr;
-		len = req->buf[1].len;
+	struct virtio_net_dev *net_dev = netdev_of(dev);
+
+	/* Let's try the first buffer first */
+	strip_dev_buf_header(&req->buf[0], &out_buf, &header);
+
+	/* The first buffer was just a header. If this request has
+	 * multiple buffers, it's a TX packet and the info we care
+	 * about is in the second buffer (because an empty packet is
+	 * useless to the TAP device), so let's send that. If we don't
+	 * have multiple buffers, this must be an RX packet, and its
```

Comment:
@stfairy Thanks for the correction! Honestly, I stared at `if (!len) { buf = &header[1] }` every day for a month before I figured out what was going on there. I hope @opurdila can point out if there's anything else to be corrected with this. It seems like this code is very tricky because most of the info about what we're trying to do has been lost by this point. I'd like to add in comments + error checks to reduce the burden on future maintainers.

My understanding now is:

* RX: `len == 0 && buf_count == 1` --> normal case -- we've just stuck a `virtio_net` header on top of a packet we're about to read data into
* RX: `len == 0 && buf_count > 1` --> error (?) the second buffer will be ignored
* RX: `len > 0` --> error (?) this will be clobbered by whatever we read in


* TX: `len > 0 && buf_count == 1` --> normal case, send on the data from that one packet
* TX: `len == 0 && buf_count == 2` --> we've ended up with 1 bogus packet because of weird fragmentation that happened somewhere else, ignore the bogus packet and send on the good packet (or is this an error?)
* TX: `len == 0 && buf_count > 2` --> error (?) we'll drop a buffer after the second one
* TX: `len > 0 && buf_count > 1` --> error (?) we'll drop a buffer after the first one

I'll update the comments on Monday and add in `WARN_ON` statements to handle what I think are error cases unless I'm overlooking something.

(Also, FWIW, I've compared the value of `buf` at this step against the output from `tcpdump`, and it looks like it would be sufficient to drop that (+ a tiny bit of metadata) into a file to get valid `pcap` data -- I dropped this idea because it turned out not to be the solution to the problem we were facing, but it would be trivial to stick in an `LKL_HIJACK_TCPDUMP_TO_FILE` environment variable if it's needed down the line.)